// @microsoft/teamsfx-core/build/component/driver/file/createOrUpdateJsonFile.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateOrUpdateJsonFileDriver = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs-extra"));
const typedi_1 = require("typedi");
const lib_1 = require("@feathersjs/hooks/lib");
const teamsfx_api_1 = require("@microsoft/teamsfx-api");
const localizeUtils_1 = require("../../../common/localizeUtils");
const common_1 = require("../../utils/common");
const constants_1 = require("../aad/utility/constants");
const addStartAndEndTelemetry_1 = require("../middleware/addStartAndEndTelemetry");
const updateProgress_1 = require("../middleware/updateProgress");
const common_2 = require("../../../error/common");
const actionName = "file/createOrUpdateJsonFile";
const helpLink = "https://aka.ms/teamsfx-actions/file-createOrUpdateJsonFile";
let CreateOrUpdateJsonFileDriver = class CreateOrUpdateJsonFileDriver {
    constructor() {
        this.description = localizeUtils_1.getLocalizedString("driver.file.createOrUpdateJsonFile.description");
    }
    async run(args, context) {
        return common_1.wrapRun(async () => {
            const result = await this.handler(args, context);
            return result.output;
        });
    }
    async execute(args, ctx) {
        let summaries = [];
        const outputResult = await common_1.wrapRun(async () => {
            const result = await this.handler(args, ctx);
            summaries = result.summaries;
            return result.output;
        });
        return {
            result: outputResult,
            summaries,
        };
    }
    async handler(args, context) {
        var _a, _b;
        try {
            this.validateArgs(args);
            const appsettingsPath = common_1.getAbsolutePath(args.target, context.projectPath);
            if (!(await fs.pathExists(appsettingsPath))) {
                // try to copy appsettings.json
                const appsettingsTemplatePath = common_1.getAbsolutePath("appsettings.json", context.projectPath);
                if (!fs.existsSync(appsettingsTemplatePath)) {
                    await fs.writeFile(appsettingsPath, "{}");
                }
                else {
                    await fs.copyFile(appsettingsTemplatePath, appsettingsPath);
                }
            }
            const appSettingsJson = JSON.parse(fs.readFileSync(appsettingsPath, "utf-8"));
            this.replaceProjectAppsettings(appSettingsJson, args.appsettings);
            await fs.writeFile(appsettingsPath, JSON.stringify(appSettingsJson, null, "\t"), "utf-8");
            return {
                output: new Map(),
                summaries: [localizeUtils_1.getLocalizedString("driver.file.createOrUpdateJsonFile.summary", args.target)],
            };
        }
        catch (error) {
            if (error instanceof teamsfx_api_1.UserError || error instanceof teamsfx_api_1.SystemError) {
                (_a = context.logProvider) === null || _a === void 0 ? void 0 : _a.error(localizeUtils_1.getLocalizedString(constants_1.logMessageKeys.failExecuteDriver, actionName, error.displayMessage));
                throw error;
            }
            const message = JSON.stringify(error);
            (_b = context.logProvider) === null || _b === void 0 ? void 0 : _b.error(localizeUtils_1.getLocalizedString(constants_1.logMessageKeys.failExecuteDriver, actionName, message));
            throw new common_2.UnhandledError(error, actionName);
        }
    }
    validateArgs(args) {
        const invalidParameters = [];
        if (args.target === undefined) {
            invalidParameters.push("target");
        }
        else if (args.target !== undefined &&
            (typeof args.target !== "string" || args.target.length === 0)) {
            invalidParameters.push("target");
        }
        if (!args.appsettings || typeof args.appsettings !== "object") {
            invalidParameters.push("appsettings");
        }
        for (const value of Object.values(args.appsettings)) {
            if (!value) {
                invalidParameters.push("appsettings");
            }
        }
        if (invalidParameters.length > 0) {
            throw new common_2.InvalidActionInputError(actionName, invalidParameters, helpLink);
        }
    }
    replaceProjectAppsettings(projectAppsettings, ymlAppsettings) {
        for (const item of Object.entries(ymlAppsettings)) {
            if (typeof item[1] === "string") {
                projectAppsettings[item[0]] = item[1];
            }
            else if (typeof item[1] === "object") {
                if (projectAppsettings[item[0]]) {
                    this.replaceProjectAppsettings(projectAppsettings[item[0]], item[1]);
                }
                else {
                    projectAppsettings[item[0]] = item[1];
                }
            }
        }
    }
};
tslib_1.__decorate([
    lib_1.hooks([
        addStartAndEndTelemetry_1.addStartAndEndTelemetry(actionName, actionName),
        updateProgress_1.updateProgress(localizeUtils_1.getLocalizedString("driver.file.progressBar.appsettings")),
    ]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], CreateOrUpdateJsonFileDriver.prototype, "run", null);
tslib_1.__decorate([
    lib_1.hooks([
        addStartAndEndTelemetry_1.addStartAndEndTelemetry(actionName, actionName),
        updateProgress_1.updateProgress(localizeUtils_1.getLocalizedString("driver.file.progressBar.appsettings")),
    ]),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object, Object]),
    tslib_1.__metadata("design:returntype", Promise)
], CreateOrUpdateJsonFileDriver.prototype, "execute", null);
CreateOrUpdateJsonFileDriver = tslib_1.__decorate([
    typedi_1.Service(actionName) // DO NOT MODIFY the service name
], CreateOrUpdateJsonFileDriver);
exports.CreateOrUpdateJsonFileDriver = CreateOrUpdateJsonFileDriver;
